### PR TITLE
Show stacktrace for exception other than ViewVCException if cfg.option.stacktraces=1

### PR DESCRIPTION
**Pull request by futatuki**
For development and debugging, it is convenient if stacktrace can be shown other than ViewVCException.

This also make ViewVCException subclass of Exception (to prepare to support Python 3).

### CODE REVIEW

**Comment by cmpilato** [comment]
I confess that I don't understand what this change is about.  I mean, I get the bit about making ViewVCException derive from Exception.  But the rest is a mystery to me.  I would expect it to have the effect of _not_ showing a stacktrace for non-ViewVCExceptions, because the template logic is such that stacktraces are only shown when there's no `msg` member of the template data dictionary.

**Comment by futatuki** [comment]
I'm sorry, my explanation is bad. The second hunk in `GetExceptionData()` makes All Exception class to show its message.  I meant the word "This" in second sentence indicates whole of this PR, but not  first sentence.

**Comment by futatuki** [comment]
The aim of this PR is just I wrote to the title and first sentence of my first explanation. The change to `ViewVCException` is extra thing.

I tried to make this PR first time, it is a rusult of cherry-pick from my python3_support_with_altsvn branch which contains all changes to `lib/debug.py`.  Then I found it breaks Python < 2.6 compatibility, so I remake a branch not to break obiously, and only change to to  is remained as change not related to the main aim. This is how and why I made bad PR....

**Comment by cmpilato** [comment]
"Show stacktrace for exception other than ViewVCException if cfg.option.stacktraces=1"

ViewVC's out of the box behavior (before the `stacktraces` option was added and now when `stacktraces=1`) is to show stacktraces for all Exceptions _except_ ViewVCExceptions.  The code you've added is not going to have the effect of showing additional stacktraces.  If anything, what you are doing is _suppressing_ the display of _all_ stacktraces.  Take a look at the relevant bit of `error.ezt`:

```
[if-any msg][else]
<h4>Python Traceback</h4>
<p><pre>
[stacktrace]
</pre></p>
[end]
```

If you set the `msg` member of `exc_dict`, ViewVC will **not** show any stacktrace for that error, regardless of how the `stacktraces` configuration option is set.

Here's an example.  With the current `master` code, if I visit a URL such as `http://viewvc-master/svn/viewvc/trunk/misc/?pathrev=zzyyza`, I see:

```
An Exception Has Occurred
Python Traceback
  File "/opt/viewvc-master/lib/viewvc.py", line 5199, in main
    request.run_viewvc()
  File "/opt/viewvc-master/lib/viewvc.py", line 329, in run_viewvc
    self.pathtype = _repos_pathtype(self.repos, path_parts, pathrev)
  File "/opt/viewvc-master/lib/viewvc.py", line 793, in _repos_pathtype
    return repos.itemtype(path_parts, rev)
  File "/opt/viewvc-master/lib/vclib/svn/svn_repos.py", line 391, in itemtype
    rev = self._getrev(rev)
  File "/opt/viewvc-master/lib/vclib/svn/svn_repos.py", line 830, in _getrev
    raise vclib.InvalidRevision(rev)
InvalidRevision: Invalid revision zzyyza
```

But viewing the same URL with your code in place, the stack trace actually disappears from the display and is replaced with the "human-friendly" error message:

```
An Exception Has Occurred
Invalid revision zzyyza

HTTP Response Status
500 Internal Server Error
```

This seems like exactly the opposite of what you're trying to accomplish.

To summarize:

By default, ViewVC shows stack traces for all exceptions that are _not_ ViewVCExceptions when `stacktraces=1`.  The only way to see _more_ stack traces would be to also show the stack traces for ViewVCExceptions, too.  Since you'd only want to do this for development and debugging, the way to do so is to replace `[if-any msg][else]` in `error.ezt` with `[if-any stacktrace]`.

**Comment by futatuki** [comment]
Ah, I understand.   You are right. I had tested mixed revision tree, so I confused. 
I'm very sorry to bother you.

**Comment by cmpilato** [comment]
No bother at all!  Happy to have somebody besides only myself paying attention to ViewVC.
